flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -290,17 +290,33 @@ private void registerCallback() {
     @Override
     public void write(InputT element, Context context) throws IOException, InterruptedException {
         while (bufferedRequestEntries.size() >= maxBufferedRequests) {
-            mailboxExecutor.tryYield();
+            flush();
         }
 
         addEntryToBuffer(elementConverter.apply(element, context), false);
 
-        flushIfAble();
+        nonBlockingFlush();
     }
 
-    private void flushIfAble() {
-        while (bufferedRequestEntries.size() >= maxBatchSize
-                || bufferedRequestEntriesTotalSizeInBytes >= maxBatchSizeInBytes) {
+    /**
+     * Determines if a call to flush will be non-blocking (i.e. {@code inFlightRequestsCount} is
+     * strictly smaller than {@code maxInFlightRequests}). Also requires one of the following
+     * requirements to be met:
+     *
+     * <ul>
+     *   <li>The number of elements buffered is greater than or equal to the {@code maxBatchSize}
+     *   <li>The sum of the size in bytes of all records in the buffer is greater than or equal to
+     *       {@code maxBatchSizeInBytes}
+     * </ul>
+     */
+    private void nonBlockingFlush() {
+        boolean uncompletedInFlightResponses = true;
+        while (uncompletedInFlightResponses) {
+            uncompletedInFlightResponses = mailboxExecutor.tryYield();
+        }
```

Comment:
6. To that end, these lines simply completes any in flight requests that have completed asynchronously and called the `completeRequest` method, thereby reducing the in flight request count. This loop will not block and will return immediately if there are none to be completed. 

7. This then allows the next method to accurately determine whether the call to flush will be blocking. If it will not be blocking, we benefit from:
 * eagerly flushing a batch of records that have reached our desired size (by number or by bytes), and,
 * failed elements will be retried sooner than otherwise due to eager completion of in flight requests.
If it will be blocking, we benefit from:
 * returning immediately and not calling flush - and therefore not unnecessarily blocking because the element has been already added to the buffer.

Reply:
I would remove this loop. It adds complexity and a overhead of checking the mailbox queue, while it doesn't do much. Any mails if present, should have been executed before any `write()` call has been made. There is usually very very small time window for a race condition to happen that task thread decided to process a new record, but before we reached this place, some batch has completed. If this happens, we can always call `nonBlockingFlush()` method from the code that is executed at the end of completing a batch (`AsyncSinkWriter#completeRequest` ?) - which we should be doing regardless of anything else.

For example what if we have just written a last record before a long pause. We haven't flushed that record, because the call would have been blocking, but it should have been flushed, because max buffered size in bytes has been reached. In this scenario after completing a batch, we should check if we should flush a new one.

Reply:
I think this is completely wonderful. I believe then the code for `completeRequest()` should be:

```
    private void completeRequest(List<RequestEntryT> failedRequestEntries, long requestStartTime) {
        // do completeRequest stuff including reducing the inFlightRequestsCount, etc.
        mailboxExecutor.tryYield();
        nonBlockingFlush();
    }
```

We would get all the previously alluded to benefits.

My only question (and maybe worry) would be if we had a very large number of in flight requests that have all completed (say 100+? since it's customisable by user). Once one `completeRequest` is triggered, the `mailboxExecutor.tryYield()` would repeatedly yield to the next completed request in a daisy chain of length equal to the number of completed in flight requests. 

I imagine the state of the mailbox thread would have to live alongside the others' states during that time and I was wondering would that risk us using more stack/heap than otherwise and thereby risking an overflow?

Reply:
I think the code should look more like this:
```
    private void completeRequest(List<RequestEntryT> failedRequestEntries, long requestStartTime) {
        lastSendTimestamp = requestStartTime;
        ackTime = System.currentTimeMillis();
        inFlightRequestsCount--;
        ListIterator<RequestEntryT> iterator =
                failedRequestEntries.listIterator(failedRequestEntries.size());
        while (iterator.hasPrevious()) {
            addEntryToBuffer(iterator.previous(), true);
        }

        nonBlockingFlush();
    }
```
Without the `tryYield()`. When executing `completeRequest()` we are already in the mail. If there are more mails coming after this `completeRequest()`, they will be executed independently of `tryYield()` call anyway. There is no need to chain those mails together.

Reply:
I agree. Thank you. This is wonderful.

Reply:
Just one last question:

You previously mentioned that:

"Any mails if present, should have been executed before any write() call has been made."

Does this mean that there is some mechanism in the sink operator that will yield to the mailbox executor to finish off completed requests in addition to `mailboxExecutor.yield();` in the `flush()` method? 

Specifically, does it call `mailboxExecutor.tryYield();` repeatedly similar to what I was doing, before executing `write()`?

Reply:
There is something like that, but not on the operator level. It's on the task level. Task thread is (apart of initialisation and closing sequences) executing always `StreamTask#runMailboxLoop`. Which is effectively:
```
        while (isNextLoopPossible()) {
            // The blocking `processMail` call will not return until default action is available.
            processMail(localMailbox, false);
            if (isNextLoopPossible()) {
                mailboxDefaultAction.runDefaultAction(
                        defaultActionContext); // lock is acquired inside default action as needed
            }
        }
```
And `mailboxDefaultAction.runDefaultAction(...)` equates to `StreamTask#processInput`. StreamTask#processInput` polls a single record from the input (source/network) and pushes it through the `OperatorChain`, which at the end means `AsyncSinkWriter#write()`. So in other words, all known/existing mails are always executed before any record is processed.